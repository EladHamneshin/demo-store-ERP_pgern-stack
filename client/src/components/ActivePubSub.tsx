import { useSubscription } from "@apollo/client";
import { NEWS_SUBSCRIPTION } from "../graphQLQueries/subscriptions";

