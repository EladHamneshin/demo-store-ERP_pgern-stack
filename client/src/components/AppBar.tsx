import { MouseEvent, useState } from 'react';
import {
  AppBar as MUIAppBar,
  Toolbar,
  IconButton,
  Typography,
  Box,
  Tooltip,
  Avatar,
  Menu,
  MenuItem,
  ListItemIcon,
} from '@mui/material';
import StorefrontIcon from '@mui/icons-material/Storefront';
import { useNavigate } from 'react-router-dom';
import Logout from '@mui/icons-material/Logout';
import SearchField from './SearchField';
import usersAPI from '../api/userAPI.ts';
import ROUTES from '../routes/routes.ts';
import { toastError, toastSuccess } from '../utils/toastUtils.ts';


const AppBar = () => {
  const [anchorElUser, setAnchorElUser] = useState<null | HTMLElement>(null);
  const navigate = useNavigate();

  const handleOpenUserMenu = (event: MouseEvent<HTMLElement>) => {
    setAnchorElUser(event.currentTarget);
  };

  const handleCloseUserMenu = () => {
    setAnchorElUser(null);
  };

  const handleLogout = async () => {
    handleCloseUserMenu();
    try {
      await usersAPI.logoutUser();
      navigate(ROUTES.LOGIN);
      toastSuccess('User logged out successfully');
    } catch (err) {
      toastError((err as Error).message);
    }
  };

  return (
    <MUIAppBar position="static" color='primary'>
      <Toolbar sx={{ display: 'flex', justifyContent: 'space-between' }}>
        <Box
          onClick={() => navigate('/')}
          sx={{
            display: 'flex',
            justifyContent: 'space-between',
            alignItems: 'center',
            cursor: 'pointer',
          }}
        >
          <StorefrontIcon sx={{ marginRight: 2 }} />
          <Typography variant="h6" component="div" sx={{ marginRight: 2 }}>
            Demo Store Inventory
          </Typography>
          <SearchField />
        </Box>
        <Box sx={{ display: 'flex', justifyContent: 'space-between' }}>
          <Box sx={{ flexGrow: 0 }}>
            <Tooltip title="Open settings">
              <IconButton onClick={handleOpenUserMenu} sx={{ p: 0 }}>
                <Avatar src="" />
              </IconButton>
            </Tooltip>
            <Menu
              sx={{ mt: '45px' }}
              id="menu-appbar"
              anchorEl={anchorElUser}
              anchorOrigin={{
                vertical: 'top',
                horizontal: 'right',
              }}
              keepMounted
              transformOrigin={{
                vertical: 'top',
                horizontal: 'right',
              }}
              open={Boolean(anchorElUser)}
              onClose={handleCloseUserMenu}
            >
              <MenuItem key={'email'} disabled>
                <Typography>test</Typography>
              </MenuItem>
              <MenuItem key={'logOut'} onClick={handleLogout}>
                <ListItemIcon>
                  <Logout fontSize="small" />
                </ListItemIcon>
                Logout
              </MenuItem>
            </Menu>
          </Box>
        </Box>
      </Toolbar>
    </MUIAppBar>
  );
};

export default AppBar;
