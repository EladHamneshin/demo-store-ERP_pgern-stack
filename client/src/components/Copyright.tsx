import Typography from '@mui/material/Typography';

export function Copyright(props: any) {
    return (
      <Typography variant="body2" color="text.secondary" align="center" {...props}>
        {'Copyright Â© '}
        {'ERP Website'}{' '}
        {new Date().getFullYear()}
        {'.'}
      </Typography>
    );
}