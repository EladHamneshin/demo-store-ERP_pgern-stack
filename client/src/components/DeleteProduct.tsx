import React, { useState } from 'react';
import { Delete } from '@mui/icons-material';
import { Button, Dialog, DialogTitle, DialogContent, DialogActions } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import productsAPI from '../api/productsAPI';
import ROUTES from '../routes/routes';

type Props = {
  id: string;
};

function DeleteProduct(props: Props) {
  const navigate = useNavigate();
  const [open, setOpen] = useState(false);

  const handleOpen = () => setOpen(true);
  const handleClose = () => setOpen(false);

  const onDelete = async () => {
    try {
      // Delete the product
      await productsAPI.deleteProduct(props.id);
      navigate(ROUTES.HOME);
    } catch (error) {
      console.error('Error deleting product:', error);
    } finally {
      handleClose();
    }
  };

  return (
    <>
      <Button variant="contained" color="error" endIcon={<Delete />} onClick={handleOpen}>
        Delete
      </Button>

      <Dialog open={open} onClose={handleClose}>
        <DialogTitle>Confirm Deletion</DialogTitle>
        <DialogContent>
          Are you sure you want to delete this product?
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>Cancel</Button>
          <Button onClick={onDelete} color="error">
            Confirm Delete
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
}

export default DeleteProduct;
