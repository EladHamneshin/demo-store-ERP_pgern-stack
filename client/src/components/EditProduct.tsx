import { Category, Product } from '../types/Product';
import { Edit } from '@mui/icons-material';
import {
  Button,
  TextField,
  Container,
  Box,
  Autocomplete,
  Switch,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle
} from '@mui/material';
import * as React from 'react';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { useNavigate } from 'react-router-dom';
import ROUTES from '../routes/routes';
import productsAPI from '../api/productsAPI';

// TODO remove, this demo shouldn't need to reset the theme.
const defaultTheme = createTheme();

type Props = {
  product: Product
}

function UpdateProduct(props: Props) {
  const [product, setProduct] = React.useState(props.product)
  const [name, setName] = React.useState(product.name);
  const [saleprice, setSalePrice] = React.useState(product.saleprice);
  const [quantity, setQuantity] = React.useState(product.quantity);
  const [description, setDescription] = React.useState(product.description);
  const [category, setCategory] = React.useState(product.category);
  const [discount, setDiscount] = React.useState(product.discount);
  const [rating, setRating] = React.useState(product.rating);
  const [clicked, setClick] = React.useState(product.clicked);
  const [url, setUrl] = React.useState(product.image.url);
  const [alt, setAlt] = React.useState(product.image.alt);
  const [isforsale, setIsForSale] = React.useState(product.isforsale);
  const [longitude, setLongitude] = React.useState(product.coordinate.longitude);
  const [latitude, setLatitude] = React.useState(product.coordinate.latitude);
  const [costprice, setCostPrice] = React.useState(product.costPrice);
  const [supplier, setSupplier] = React.useState(product.supplier);
  const [tags, setTags] = React.useState({ ...product.tags });
  const [categories, setCategories] = React.useState<Category[]>([{name: 'foo', id: '1', clicked: 0}, {name: 'bar', id: '2', clicked: 0}])

  const [open, setOpen] = React.useState(false); //for editing the product
  const navigate = useNavigate();
  const handleOpen = async() => {
    const categories = await productsAPI.getCategories();
    console.log(categories);
    
    setOpen(true);
    setCategories(categories)};
  const handleClose = () => {
    setOpen(false);
  };

  React.useEffect(() => {
    if (!open) {
      // Reset the state to the initial values when the dialog is closed
      setProduct(product);
      setName(product.name);
      setSalePrice(product.saleprice);
      setQuantity(product.quantity);
      setDescription(product.description);
      setCategory(product.category);
      setDiscount(product.discount);
      setRating(product.rating);
      setClick(product.clicked);
      setUrl(product.image.url);
      setAlt(product.image.alt);
      setIsForSale(product.isforsale);
      setLongitude(product.coordinate.longitude);
      setLatitude(product.coordinate.latitude);
      setCostPrice(product.costPrice);
      setSupplier(product.supplier);
      setTags({ ...product.tags });
    }
  }, [open, product]);

  const submitUpdates = async () => {
    const updatedProduct: Product = {
      id: product.id,
      category,
      clicked,
      description,
      coordinate: {
        latitude,
        longitude
      },
      costPrice: costprice,
      image: {
        alt,
        url
      },
      supplier,
      isforsale: isforsale,
      name,
      quantity,
      rating,
      saleprice,
      tags: { ...tags },
      discount
    };
    try {
      const req = await productsAPI.updateProduct(updatedProduct, product.id!);
      console.log('ererer', req);
      console.log('Updated Product');
    } catch (err) {
      console.log('failed to update Product');
    } finally {
      navigate(`${ROUTES.PRODUCT_ROUTE}/${product.id}`)

    }
  };

  const updateTag = (oldKey: string, newKey: string, value: string) => {
    const updatedTags = { ...tags };
    delete updatedTags[oldKey]; // Remove the old key
    updatedTags[newKey] = value; // Add the new key
    setTags(updatedTags);
  };

  const deleteTag = (key: string) => {
    const updatedTags = { ...tags };
    delete updatedTags[key]; // Remove the key
    setTags(updatedTags);
  }

  const addNewTag = () => {
    setTags((prevTags) => ({
      ...prevTags,
      '': '',
    }));
  };


  const isForSaleHandleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setIsForSale(event.target.checked);
  };

  return (
    <>
      <Button variant="contained" color="primary" endIcon={<Edit />} onClick={handleOpen}>
        Edit
      </Button>
      <ThemeProvider theme={defaultTheme}>
        <Dialog open={open} onClose={handleClose}>
          <Container component="main" maxWidth="xs">
            <Box
              sx={{
                marginTop: 8,
                display: 'flex',
                flexDirection: 'column',
                alignItems: 'center',
              }}
            >
              <DialogTitle>Edit Product</DialogTitle>
              <DialogContent>
                <TextField
                  margin="normal"
                  fullWidth
                  label="Name"
                  value={name}
                  onChange={(e) => { setName(e.target.value) }}
                />
                <TextField
                  type="number"
                  margin="normal"
                  fullWidth
                  label="Sale Price"
                  value={saleprice}
                  onChange={(e) => setSalePrice(Number(e.target.value))}
                />
                <TextField
                  type="number"
                  margin="normal"
                  fullWidth
                  label="Quantity"
                  value={quantity}
                  onChange={(e) => setQuantity(Number(e.target.value))}
                />
                <TextField
                  margin="normal"
                  fullWidth
                  label="Description"
                  value={description}
                  onChange={(e) => setDescription(e.target.value)}
                  multiline
                />
                <Autocomplete
                  id="Category"
                  freeSolo
                  fullWidth
                  value={category}
                  options={categories!.map((option) => option.name)}
                  renderInput={(params) => <TextField margin='normal' {...params} label="Category" />}
                  onChange={(_, newValue) => {
                    newValue ? setCategory(newValue): setCategory(category)
                  }}
                />
                <TextField
                  type="number"
                  margin="normal"
                  fullWidth
                  label="Discount"
                  value={discount}
                  onChange={(e) => setDiscount(Number(e.target.value))}
                />
                <TextField
                  type="number"
                  margin="normal"
                  fullWidth
                  label="Rating"
                  value={rating}
                  onChange={(e) => setRating(Number(e.target.value))}
                />
                <TextField
                  type="number"
                  margin="normal"
                  fullWidth
                  label="Click"
                  value={clicked}
                  onChange={(e) => setClick(Number(e.target.value))}
                />
                <TextField
                  margin="normal"
                  fullWidth
                  label="Image URL"
                  value={url}
                  onChange={(e) => setUrl(e.target.value)}
                />
                <TextField
                  margin="normal"
                  fullWidth
                  label="Image Alt"
                  value={alt}
                  onChange={(e) => setAlt(e.target.value)}
                />
                <TextField
                  type="number"
                  margin="normal"
                  fullWidth
                  label="Longitude"
                  value={longitude}
                  onChange={(e) => setLongitude(Number(e.target.value))}
                />
                <TextField
                  type="number"
                  margin="normal"
                  fullWidth
                  label="Latitude"
                  value={latitude}
                  onChange={(e) => setLatitude(Number(e.target.value))}
                />
                <TextField
                  type="number"
                  margin="normal"
                  fullWidth
                  label="Cost Price"
                  value={costprice}
                  onChange={(e) => setCostPrice(Number(e.target.value))}
                />
                <TextField
                  margin="normal"
                  fullWidth
                  label="Supplier"
                  value={supplier}
                  onChange={(e) => setSupplier(e.target.value)}
                />
                <h3>Tags</h3>
                {Object.keys(tags).map((key, index) => (
                  <div key={index} style={{ display: "flex" }}>
                    <TextField
                      margin="normal"
                      label="Tag Name"
                      value={key}
                      onChange={(e) => updateTag(key, e.target.value, tags[key])}
                    />
                    <TextField
                      margin="normal"
                      label="Tag Value"
                      value={tags[key]}
                      onChange={(e) => updateTag(key, key, e.target.value)}
                    />
                    <Button onClick={() => deleteTag(key)}>Delete</Button>
                  </div>
                ))}
                <Button onClick={addNewTag}>Add Tag</Button>
                <span>Is For Sale</span>
                <Switch
                  checked={isforsale}
                  onChange={isForSaleHandleChange}
                  inputProps={{ 'aria-label': 'controlled' }}
                />
              </DialogContent>
              <DialogActions>
                <Button onClick={handleClose}>Cancel</Button>
                <Button onClick={submitUpdates}>Update Product</Button>
              </DialogActions>
            </Box>
          </Container>
        </Dialog>
      </ThemeProvider>
    </>
  )
}

export default UpdateProduct