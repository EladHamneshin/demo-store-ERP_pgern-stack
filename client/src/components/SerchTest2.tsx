import SearchIcon from '@mui/icons-material/Search';
import Autocomplete from '@mui/material/Autocomplete';
import { styled, alpha, InputBase, TextField } from '@mui/material';
import { useState, useEffect } from 'react';
import productsAPI from '../api/productsAPI';
import { Product } from '../types/Product';
import { useNavigate } from 'react-router-dom';

const Search = styled('div')(({ theme }) => ({
  position: 'relative',
  borderRadius: theme.shape.borderRadius,
  backgroundColor: alpha(theme.palette.common.white, 0.15),
  '&:hover': {
    backgroundColor: alpha(theme.palette.common.white, 0.25),
  },
  marginRight: theme.spacing(2),
  marginLeft: 0,
  width: '100%',
  [theme.breakpoints.up('sm')]: {
    marginLeft: theme.spacing(3),
    width: 'auto',
  },
}));

const SearchIconWrapper = styled('div')(({ theme }) => ({
  padding: theme.spacing(0, 2),
  height: '100%',
  position: 'absolute',
  pointerEvents: 'none',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
}));

export default function SearchTest2() {
  const navigate = useNavigate();
  const [searchRes, setSearchRes] = useState<Product[]>([]);

  useEffect(() => {
    const getProducts = async () => {
      const allProduct = await productsAPI.getAllProducts();
      setSearchRes(allProduct);
    }
    getProducts()
  }, []);

  const handleSelectOption = (event: React.ChangeEvent<{}>, value: string | null) => {
    if (value) {
      const selectedProduct = searchRes.find(product => product.name === value);
      if (selectedProduct) {
        navigate(`/erp/product/${selectedProduct.id}`);
        console.log('id:', selectedProduct.id);
        console.log('url: ', `/erp/product/${selectedProduct.id}`);
      }
    }
  }

  return (
    <Search>
      <Autocomplete
        freeSolo
        id="free-solo-2-demo"
        disableClearable
        options={searchRes.map((option) => option.name)}
        onChange={handleSelectOption}
        renderInput={(params) => (
          <TextField
            {...params}
            placeholder="Searchâ€¦"
            // label="Search input"
            InputProps={{
              ...params.InputProps,
              type: 'search',
            }}
            sx={{ width: 300 }}
          />
        )}
      />
    </Search>
  )
}