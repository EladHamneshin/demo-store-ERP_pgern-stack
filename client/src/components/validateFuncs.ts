export const passwordValidet = {
    required: "⚠ Required field",
    pattern: {
      value: /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*]).{8,20}$/,
      message:
        "⚠ Password must contain an uppercase letter, a lowercase letter, a digit, a special character, and be 8-20 characters long.",
    },
  };
  export const emailValidet = {
    required: "⚠ Required field",
    pattern: {
      value: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
      message: "⚠ Invalid email address",
    },
  };
  export const nameValidet = {
    required: "⚠ Required field",
    minLength: {
      value: 2,
      message: "⚠ Must be at least two characters",
    },
  };

// function isValidEmail(email: string) {
//     return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
// }

// function isValidPassword(password: string) {
//     return /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+=!])(?=.{7,20}$)/.test(password);
// }

// function isValidName(name: string){
//     return name !== '';
// }

// export { isValidEmail, isValidPassword, isValidName };


