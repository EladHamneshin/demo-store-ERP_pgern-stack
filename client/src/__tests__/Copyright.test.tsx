import { render, screen } from '@testing-library/react';
import { describe, it, expect } from 'vitest';

import { Copyright } from '../components/Copyright';

describe('Copyright', () => {

  it('renders the current year', () => {

    render(<Copyright />);
    

    expect(screen.getByText(/2024/)).toBeInTheDocument();

  });

  it('renders the copyright symbol', () => {

    render(<Copyright />);
    
    expect(screen.getByText(/Â©/)).toBeInTheDocument();

  });

  it('matches snapshot', () => {

    const { container } = render(<Copyright />);
    
    expect(container).toMatchSnapshot();

  });

});