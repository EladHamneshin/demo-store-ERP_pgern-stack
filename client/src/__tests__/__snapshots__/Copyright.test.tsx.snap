// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Copyright > matches snapshot 1`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter css-1vl95ex-MuiTypography-root"
  >
    Copyright Â© 
    ERP Website
     
    2024
    .
  </p>
</div>
`;
